### DISCUSSION TITLE
$patch function replace value in remote object

### DISCUSSION THREAD

**Discussion by mathias22osterhagen22** [vuejs/pinia#1921] [Help and Questions] [answered]
### Reproduction

https://github.com/mathias22osterhagen22/vuejs-pinia-patch-issue.git

### Steps to reproduce the bug

1. Create an Enum with object as keys (or any object)
![image](https://user-images.githubusercontent.com/60438474/211343890-d8230a2f-70d2-4be4-81ee-0c0ea7933ef2.png)
2. Create a property in the state & assign a value from the enum object (or a prop of an object)
![image](https://user-images.githubusercontent.com/60438474/211344032-80b8332d-0921-429c-8f2f-a324e73b4771.png)
3. Patch the property of the state with an object having the same properties
![image](https://user-images.githubusercontent.com/60438474/211344132-0933ba06-2a26-4d7e-9414-36b891dc5a1e.png)

### Expected behavior

1. State updating ✅
2. Object values updating ❌

### Actual behavior

1. State updating ✅
2. Object values updating ✅ <--- this should not happend

### Additional information

Object before patch: 
![image](https://user-images.githubusercontent.com/60438474/211344574-cdf91b89-d3e6-4fb4-bb2c-c0f7cef1fcc5.png)

Patching "TEST" updating the state + the object values passed as ref: 
![image](https://user-images.githubusercontent.com/60438474/211344732-38edc7f4-9758-4da2-95d3-41e14a597b3d.png)

This seems to happened because the `$patch` method is updating from bottom to top recursively: 
https://github.com/vuejs/pinia/blob/78ec9a186dcbce3d583db332ae22094a182358cc/packages/pinia/src/store.ts#L83

I met this issue because of the Local storage "loading" (with the plugin [https://github.com/prazdevs/pinia-plugin-persistedstate]( https://github.com/prazdevs/pinia-plugin-persistedstate))

I'm storing an Enum object in the state, stored in the local storage. On page reload the  `$patch` function is called and come to replace the value Enum object value for the entire website.

**Comment by LinusBorg** [comment, marked as answer]
Assigning an object to a different variable/object property does not make a copy. it just makes a new reference to the same object, so changing one changes the "other" as well.

That's just how JS works.

**Reply by mathias22osterhagen22** [reply]
@LinusBorg  You didn't understand, I known how "JS works", the problem is that `$patch` assign values recursively.

In this case: 
```js
let a = {age: 12, name: 'test'};
let b = {age: 23, name: 'second'};
let d = a; //the initial state
```
Instead of doing :
```js
d = b;

// -- RESULT --
//a = {age: 12, name: 'test'}
//b = {age: 23, name: 'second'}
//d = {age: 23, name: 'second'} (pointing on b)
```

`$patch` is doing this: 
```js
d.age = b.age;
d.name = b.name;
d = b;

// -- RESULT --
//a = {age: 23, name: 'second'} <--- 'a' object was modified because of patch
//b = {age: 23, name: 'second'}
//d = {age: 23, name: 'second'} (pointing on b)
```

As you cant see in the code: 
[https://github.com/prazdevs/pinia-plugin-persistedstate](https://github.com/prazdevs/pinia-plugin-persistedstate)
![image](https://user-images.githubusercontent.com/60438474/211358267-b1c052b7-ba8c-4bc9-b50b-ec5ac9492acb.png)

**Reply by LinusBorg** [reply]
You are right in that I misunderstood the point of your issue. However, as you correctly show in the source, that is the intended behavior of $patch and thus not a bug.

But likely something that needs a clarification in the docs @posva

**Reply by posva** [reply]
Added to the docs roadmap!
BTW, related issue:
- https://github.com/vuejs/pinia/issues/528

**Reply by mathias22osterhagen22** [reply]
Ok so for now I freeze my object (as it's an enum), but exepect just the reference to update, I don't know if it's to cover some specific points or what? Btw, seems that `$patch` is adding properties as well in the remote object.
```js
let a = {age: 12, name: 'test'};
let b = {age: 23, name: 'second', newprop: 'new'};
let d = a; //the initial state
```

`$patch`
```js
d.age = b.age;
d.name = b.name;
d.newprop = b.newprop;
d = b;

// -- RESULT --
//a = {age: 23, name: 'second', newprop: 'new'} <--- 'a' object was modified because of patch + new prop
//b = {age: 23, name: 'second', newprop: 'new'}
//d = {age: 23, name: 'second', newprop: 'new'} (pointing on b)
```

Is there any other function that do not update the referenced object subprop but just the root reference ?
Basically: 
```js
d = b;

// -- RESULT --
//a = {age: 12, name: 'test'}} <--- a not changing
//b = {age: 23, name: 'second', newprop: 'new'}
//d = {age: 23, name: 'second', newprop: 'new'} (pointing on b)
```

**_EDIT:_ @posva Could you let me know what is the cleanest way to deal with it ?** 
- Using JSON.parse(JSON.stringify(object)) to avoid the real object to be updated?
- Using Freeze on the properties of the object and the object itself to avoid it to be updated?

Thank you very much.

**Reply by LinusBorg** [reply]
```ts
$patch(state => {
state.enumValue = { /* new object*/ }
})
```
